File: frontend/lib/models/sale.dart
import 'customer.dart';
import 'squid_type.dart';

class Sale {
  final int id;
  final int customerId;
  final int squidTypeId;
  final double weight;
  final double unitPrice;
  final double totalAmount;
  final DateTime saleDate;
  final String paymentStatus;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;
  final Customer? customer;
  final SquidType? squidType;

  Sale({
    required this.id,
    required this.customerId,
    required this.squidTypeId,
    required this.weight,
    required this.unitPrice,
    required this.totalAmount,
    required this.saleDate,
    required this.paymentStatus,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
    this.customer,
    this.squidType,
  });

  factory Sale.fromJson(Map<String, dynamic> json) {
    return Sale(
      id: json['id'],
      customerId: json['customer_id'],
      squidTypeId: json['squid_type_id'],
      weight: double.parse(json['weight'].toString()),
      unitPrice: double.parse(json['unit_price'].toString()),
      totalAmount: double.parse(json['total_amount'].toString()),
      saleDate: DateTime.parse(json['sale_date']),
      paymentStatus: json['payment_status'],
      notes: json['notes'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      customer: json['customer'] != null ? Customer.fromJson(json['customer']) : null,
      squidType: json['squid_type'] != null ? SquidType.fromJson(json['squid_type']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'customer_id': customerId,
      'squid_type_id': squidTypeId,
      'weight': weight,
      'unit_price': unitPrice,
      'total_amount': totalAmount,
      'sale_date': saleDate.toIso8601String().split('T')[0],
      'payment_status': paymentStatus,
      'notes': notes,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  bool get isPaid => paymentStatus == 'paid';
  bool get isUnpaid => paymentStatus == 'unpaid';
} 

File: backend/app/Models/Sale.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Sale extends Model
{
    use HasFactory;

    protected $fillable = [
        'farm_id',
        'customer_id',
        'squid_type_id',
        'weight',
        'unit_price',
        'total_amount',
        'sale_date',
        'payment_status',
        'notes',
    ];

    protected $casts = [
        'sale_date' => 'date',
        'weight' => 'decimal:2',
        'unit_price' => 'decimal:2',
        'total_amount' => 'decimal:2',
    ];

    /**
     * Get the farm that owns the sale.
     */
    public function farm()
    {
        return $this->belongsTo(Farm::class);
    }

    /**
     * Get the customer that owns the sale.
     */
    public function customer()
    {
        return $this->belongsTo(Customer::class);
    }

    /**
     * Get the squid type that owns the sale.
     */
    public function squidType()
    {
        return $this->belongsTo(SquidType::class);
    }

    /**
     * Calculate total amount automatically
     */
    protected static function boot()
    {
        parent::boot();

        static::saving(function ($model) {
            $model->total_amount = $model->weight * $model->unit_price;
        });
    }

    /**
     * Scope a query to only include unpaid sales.
     */
    public function scopeUnpaid($query)
    {
        return $query->where('payment_status', 'unpaid');
    }

    /**
     * Scope a query to only include paid sales.
     */
    public function scopePaid($query)
    {
        return $query->where('payment_status', 'paid');
    }
}
